dayahead-optim and naive-mpc fail with "Cannot subtract tz-naive and tz-aware datetime-like objects" when using mlforecaster_weather_features

## Description

I encountered an issue related to `mlforecaster_weather_features`.

When weather covariates are enabled, model training and prediction complete successfully, but both optimization endpoints fail with a timezone-related exception.

When `mlforecaster_weather_features` is an empty list (`[]`), everything works correctly.

## Environment

- EMHASS: 0.17.9
- Home Assistant OS Add-on
- Python: 3.12.13

## Configuration

Load forecast method:

```json
"load_forecast_method": "mlforecaster",
"weather_forecast_method": "open-meteo",
"mlforecaster_weather_features": [
    "temp_air"
]
```

Everything else is left unchanged.

## Steps to reproduce

1. Set:

```json
"mlforecaster_weather_features": [
    "temp_air"
]
```

2. Run:

```
POST /action/forecast-model-fit
```

→ succeeds.

3. Run:

```
POST /action/forecast-model-predict
```

→ succeeds.

4. Run:

```
POST /action/dayahead-optim
```

or

```
POST /action/naive-mpc-optim
```

Both fail with the exception below.

If I change

```json
"mlforecaster_weather_features": []
```

and retrain the model, both optimization endpoints work again.

## Expected behavior

`dayahead-optim` and `naive-mpc-optim` should work correctly when weather covariates are enabled.

## Actual behavior

Optimization fails before solving with:

```text
TypeError: Cannot subtract tz-naive and tz-aware datetime-like objects
```

## Stack trace

```text
Traceback (most recent call last):
...
File "/app/src/emhass/forecast.py", line 978, in get_weather_covariates
    span_past_days = max(0, (now.normalize() - index.min()).days + 1)

TypeError: Cannot subtract tz-naive and tz-aware datetime-like objects.
```

## Additional notes

- `forecast-model-fit` works correctly.
- `forecast-model-predict` works correctly.
- Only optimization endpoints fail.
- Home Assistant provides location/timezone automatically (standard HAOS add-on configuration).
- The issue is reproducible every time after enabling any weather feature.

ERROR in app: Exception on request POST /action/dayahead-optim
Traceback (most recent call last):
...
File "/app/src/emhass/forecast.py", line 978, in get_weather_covariates
    span_past_days = max(0, (now.normalize() - index.min()).days + 1)

TypeError: Cannot subtract tz-naive and tz-aware datetime-like objects.

I can consistently reproduce this issue.

As soon as `mlforecaster_weather_features` is set to an empty list again and the model is retrained, both `dayahead-optim` and `naive-mpc-optim` start working normally.

﻿Had a dig into this, think I've tracked it down. Line refs below are from current master; on 0.17.9 the same code sits a bit higher in the file, but the crash line 978 is identical in both, which matches your traceback.

The crash is in `get_weather_covariates` (`forecast.py:978`):
```python
now = pd.Timestamp.now(tz=self.time_zone)
span_past_days = max(0, (now.normalize() - index.min()).days + 1)
```
`now` is always tz-aware, so this only blows up when the `index` it receives is tz-naive. That index is built in `_build_weather_future` (`forecast.py:1945`):
```python
future_index = pd.date_range(
    start=data_last_window.index[-1] + window_freq,
    periods=steps,
    freq=window_freq,
)
```
There's no `tz=` here, so the range just inherits whatever `data_last_window.index` carries. I reproduced your exact `TypeError: Cannot subtract tz-naive and tz-aware datetime-like objects` by feeding a naive start into that call. A tz-aware index in the wrong zone would not crash by the way, pandas subtracts across zones fine, I checked that too. So your index is arriving fully naive.

On why fit and predict work but the optim endpoints don't: predict actually goes through this same weather covariate code, so it's not that the optim path is special there. The difference is where `data_last_window` comes from. Fit and predict get their data from the main `RetrieveHass` object, which is constructed with your configured `time_zone`, so the index comes back tz-aware and everything works. The optim path instead builds its own `RetrieveHass` inside `_prepare_hass_load_data` with `time_zone=None` (`forecast.py:1777-1786`). Whether the data survives that with its tz intact then depends on the retrieval path: from what I can see the REST history path happens to keep a UTC-aware index, but the websocket statistics path runs the index through `tz_convert(self.time_zone)` with that `time_zone` being None (`retrieve_hass.py:1856`), and `tz_convert(None)` strips the timezone entirely. That produces exactly the naive index that crashes.

One thing you could confirm from your side, no code changes needed: do you have `use_websocket: true` in your config? If yes, that's the whole chain confirmed. If not, the naive index is coming from somewhere I haven't found yet and I'd want to dig further before claiming the full story.

For the fix, I don't think it's just the `date_range` call. I tried the obvious patch of adding `tz=self.time_zone` there and it breaks a different way when the start is already tz-aware (`AssertionError: Inferred time zone not equal to passed time zone`). The cleaner fix is the root cause: pass the configured time zone to that internal `RetrieveHass` instead of `None`, same as the other two `RetrieveHass` constructions in the codebase, which makes the optim path behave exactly like the predict path that already works. On top of that a defensive localize-or-convert on the `date_range` start (the same pattern this file already uses at `forecast.py:264-268` and `:1857-1865`) would make sure a naive index can never reach that subtraction again.

Workaround for now is what you found: `mlforecaster_weather_features: []`. That works because `_build_weather_future` returns early when the model has no weather features (`forecast.py:1932-1933`), so it just skips the broken path entirely.

Happy to put up a PR for this, it's a small self-contained change.

--- a/src/emhass/forecast.py
+++ b/src/emhass/forecast.py
@@ -1774,7 +1774,10 @@ class Forecast:
         var_list = [self.var_load]
         var_replace_zero = None
         var_interp = [self.var_load]
-        time_zone_load_forecast = None
+        # Pass the configured time zone so the retrieved index stays tz-aware: with None,
+        # the websocket statistics path runs the index through tz_convert(None), which
+        # strips the tz entirely and later crashes the weather covariate horizon build.
+        time_zone_load_forecast = self.time_zone
         rh = RetrieveHass(
             self.retrieve_hass_conf["hass_url"],
             self.retrieve_hass_conf["long_lived_token"],
@@ -1946,6 +1949,17 @@ class Forecast:
             start=data_last_window.index[-1] + window_freq,
             periods=steps,
             freq=window_freq,
+        )
+        # get_weather_covariates subtracts a tz-aware "now" from this index, so it must be
+        # tz-aware; date_range inherits tz-naivety from data_last_window's index.
+        future_index = (
+            future_index.tz_localize(
+                self.time_zone,
+                ambiguous="infer",
+                nonexistent="shift_forward",
+            )
+            if future_index.tz is None
+            else future_index.tz_convert(self.time_zone)
         )
         return await self.get_weather_covariates(future_index, weather_features)
 

--- a/tests/test_forecast.py
+++ b/tests/test_forecast.py
@@ -575,6 +575,131 @@ class TestForecast(unittest.IsolatedAsyncioTestCase):
         with self.assertRaises(ValueError, msg="Expected ValueError for non-uniform index"):
             await self.fcst._build_weather_future(data_last_window, mock_mlf)
 
+    @staticmethod
+    def _open_meteo_covariate_payload(full: pd.DatetimeIndex) -> dict:
+        """Synthetic Open-Meteo minutely_15 payload covering the ``full`` (tz-aware) index."""
+        times = (full.tz_convert("UTC").astype("int64") // 10**9).tolist()
+        n = len(full)
+        return {
+            "minutely_15": {
+                "time": times,
+                "temperature_2m": [20.0] * n,
+                "relative_humidity_2m": [50.0] * n,
+                "cloud_cover": [30.0] * n,
+                "wind_speed_10m": [5.0] * n,
+                "shortwave_radiation": [200.0] * n,
+                "direct_radiation": [150.0] * n,
+                "diffuse_radiation": [50.0] * n,
+                "precipitation": [0.0] * n,
+            }
+        }
+
+    async def test_build_weather_future_localizes_naive_window_index(self):
+        """A tz-naive last-window index no longer crashes the weather horizon build (#1036).
+
+        The optim path's websocket statistics retrieval used to hand this method a tz-naive
+        index, and the naive future horizon crashed get_weather_covariates with
+        'Cannot subtract tz-naive and tz-aware datetime-like objects'. The horizon must now
+        reach get_weather_covariates as tz-aware, in the configured time zone.
+        """
+        from unittest.mock import MagicMock, patch
+
+        num_lags = 16
+        naive_window_index = self.fcst.forecast_dates[:8].tz_localize(None)
+        data_last_window = pd.DataFrame(index=naive_window_index)
+        # Naive wall times are interpreted as the configured local time zone.
+        expected_index = pd.date_range(
+            start=naive_window_index[-1] + self.fcst.freq,
+            periods=num_lags,
+            freq=self.fcst.freq,
+        ).tz_localize(self.fcst.time_zone)
+        full = pd.date_range(
+            start=expected_index[0] - 4 * self.fcst.freq,
+            end=expected_index[-1] + 4 * self.fcst.freq,
+            freq=self.fcst.freq,
+        )
+        payload = self._open_meteo_covariate_payload(full)
+        mock_mlf = MagicMock()
+        mock_mlf.weather_features = ["temp_air"]
+        mock_mlf.is_tuned = False
+        mock_mlf.num_lags = num_lags
+
+        with patch.object(self.fcst, "_fetch_open_meteo_covariates_json", return_value=payload):
+            weather_future = await self.fcst._build_weather_future(data_last_window, mock_mlf)
+
+        self.assertIsNotNone(weather_future)
+        self.assertEqual(len(weather_future), num_lags)
+        self.assertIsNotNone(weather_future.index.tz)
+        self.assertEqual(weather_future.index.tz, self.fcst.time_zone)
+        self.assertTrue(weather_future.index.equals(expected_index))
+
+    async def test_build_weather_future_converts_aware_window_index(self):
+        """A tz-aware last-window index in another zone keeps its instants.
+
+        Cross-zone tz-aware input never crashed; the horizon is now normalized to the
+        configured time zone, which must not move the instants.
+        """
+        from unittest.mock import MagicMock, patch
+
+        num_lags = 16
+        utc_window_index = self.fcst.forecast_dates[:8].tz_convert("UTC")
+        data_last_window = pd.DataFrame(index=utc_window_index)
+        expected_index = pd.date_range(
+            start=utc_window_index[-1] + self.fcst.freq,
+            periods=num_lags,
+            freq=self.fcst.freq,
+            tz="UTC",
+        ).tz_convert(self.fcst.time_zone)
+        full = pd.date_range(
+            start=expected_index[0] - 4 * self.fcst.freq,
+            end=expected_index[-1] + 4 * self.fcst.freq,
+            freq=self.fcst.freq,
+        )
+        payload = self._open_meteo_covariate_payload(full)
+        mock_mlf = MagicMock()
+        mock_mlf.weather_features = ["temp_air"]
+        mock_mlf.is_tuned = False
+        mock_mlf.num_lags = num_lags
+
+        with patch.object(self.fcst, "_fetch_open_meteo_covariates_json", return_value=payload):
+            weather_future = await self.fcst._build_weather_future(data_last_window, mock_mlf)
+
+        self.assertIsNotNone(weather_future)
+        self.assertEqual(len(weather_future), num_lags)
+        self.assertEqual(weather_future.index.tz, self.fcst.time_zone)
+        self.assertTrue(weather_future.index.equals(expected_index))
+
+    async def test_prepare_hass_load_data_uses_configured_time_zone(self):
+        """_prepare_hass_load_data builds its internal RetrieveHass with the configured tz (#1036).
+
+        With time_zone=None the websocket statistics retrieval path strips the index tz
+        (tz_convert(None)), producing the naive index behind the #1036 crash.
+        """
+        from unittest.mock import patch
+
+        # self.fcst.get_data_from_file is True by default, so after construction the code
+        # takes the aiofiles-pickle branch, not rh.get_data() -- there's no natural await
+        # failure to stop on. Raise a sentinel from the constructor itself to capture its
+        # args without also mocking the unrelated file-read/prepare_data pipeline.
+        class _CtorCapture(Exception):
+            pass
+
+        captured = {}
+
+        def _capture_ctor(*args, **kwargs):
+            captured["args"] = args
+            captured["kwargs"] = kwargs
+            raise _CtorCapture()
+
+        with patch.object(forecast_module, "RetrieveHass", side_effect=_capture_ctor):
+            with self.assertRaises(_CtorCapture):
+                await self.fcst._prepare_hass_load_data(3, "mlforecaster")
+
+        # Signature: (hass_url, long_lived_token, freq, time_zone, params, emhass_conf, logger)
+        passed_time_zone = captured["args"][3]
+        self.assertIsNotNone(passed_time_zone)
+        self.assertEqual(passed_time_zone, self.fcst.time_zone)
+
     # Test output weather forecast using Solcast with mock get request data
     async def test_get_weather_forecast_solcast_method_mock(self):
         self.fcst.params = {

